Boosting BBVI Tutorial

lorenzkuhn opened the pull request:
`boosting_bbvi_tutorial.ipynb` contains my draft of the boosting BBVI tutorial for the Pyro documentation. Please review it - I'd be happy about any kind of feedback and suggestions. 

I used `pylint` to check for PEP8 compliance. I've prioritised consistency with existing tutorials over being compliant with PEP8 in cases where there were conflict between these two.
I've fixed the import order with `isort` as recommended in the Pyro [contribution guidelines](https://github.com/pyro-ppl/pyro/blob/dev/CONTRIBUTING.md) .

I don' think there's a need to review the other files in this PR at this point.

lorenz-openai committed:
Add comparison with standard SVI

lorenz-openai committed:
Add RELBO components plot

lorenz-openai committed:
Add first draft of BLR

lorenz-openai committed:
Fix bug in guide

lorenz-openai committed:
Use MCMC to infer weights

lorenz-openai committed:
Add debugging output

lorenz-openai committed:
Change prior; use TraceEnum_ELBO

lorenz-openai committed:
Add standard SVI as a comparison

lorenz-openai committed:
Fix tensor copying bug

lorenz-openai committed:
Evaluate model using log prob on test data

lorenz-openai committed:
Update plot of ELBO[q_t, p]

lorenz-openai committed:
Back up before refactoring guide

lorenz-openai committed:
Refactor

lorenz-openai committed:
Add BLR config that actually improves test log prob

lorenz-openai committed:
Backup best BBBVI  BLR results

lorenz-openai committed:
Add new best BLR configuration

lorenz-openai committed:
Extend BBBVI tutorial

lorenz-openai committed:
Rewrit explanations in bbbvi tutorial

lorenz-openai committed:
Add list of contents and comments

lorenz-openai committed:
Fix errors in writing

TNU-yaoy committed:
suggested changes to bbbvi tutorial

TNU-yaoy committed:
fixed syntax error

TNU-yaoy committed:
fixed syntax error

TNU-yaoy reviewed (changes requested):
I made some changes mostly to the text and committed them to a new branch (see https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/pull/14). Below are two additional comments:

1. you are using 3 different naming conventions for the latent and observed variables: z and x (intro), mu and y (model description), loc and obs (code). It might make sense to unify this if it is not too much work.

2. can we get away with not replicating the observations 10 times?

lorenzkuhn committed:
Merge pull request #14 from ratschlab/suggested_changes_tutorial

suggested changes to bbbvi tutorial

gideonite commented on boosting_bbvi_tutorial.ipynb:
```diff
@@ -0,0 +1,448 @@
+{
```
Can't give line comments in GitHub for notebooks unfortunately. Simply putting all comments here.

In the section, "The Greedy Algorithm," I would prefer if the extra stuff (imports, etc) was put somewhere else and the focus was put on the core for-loop, e.g. `for step in range(n_steps):`, and whatever is the minimal amount of things necessary to have that make sense. The full code can be included at the very end of the tutorial.

lorenzkuhn replied:
Thanks, for the review @gideonite ! 

That's a good point, I've factored out the core loop, and then put the entire implementation at the end.

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+PRINT_TRACES = False
+
+# this is for running the notebook in our testing framework
+smoke_test = ('CI' in os.environ)
```
Are we using this `smoke_test`?

lorenzkuhn replied:
The Pyro GitHub repo has a continuous integration setup where they automatically run tests whenever anything is commited. They suggest that one uses this smoke_test flag to reduce the number of iterations of large loops, s.t. the tests run faster. E.g.: `n_steps = 2 if smoke_test else 12000`

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+pyro.set_rng_seed(2)
+
+# enable validation (e.g. validate parameters of distributions)
+pyro.enable_validation(True)
```
Do we want to set this globally forever?

lorenzkuhn replied:
Thanks for taking a look at the logistic regression code as well. I've focused on the tutorial though and I haven't refactored the Bayesian Logistic Regression code yet. Will notify you when it's ready for review.

gideonite replied:
sounds good.

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+# this is for running the notebook in our testing framework
+smoke_test = ('CI' in os.environ)
+n_steps = 2 if smoke_test else 10000
+pyro.set_rng_seed(2)
```
Do we want to set this globally forever?

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+guide_log_prob = []
+approximation_log_prob = []
+
+# @config_enumerate
```
Always remove dead code.

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+
+    def get_distribution(self):
+        scale_q = pyro.param('scale_{}'.format(self.index), self.initial_scale, constraints.positive)
+        #scale_q = torch.eye(self.n_variables)
```
dead?

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+    return X_train, y_train, X_test, y_test
+
+
+def relbo(model, guide, *args, **kwargs):
```
Is there an opportunity here to factor out relbo once and for all into its own file and then use it in a modular fashion?

lorenzkuhn replied:
Yes, I think this should be possible and I got started on this. Only thing missing is to factor out the variable name that needs to be exposed when computing the trace of the approximation 
Will also be part cleaning up the BLR code.

gideonite commented on bayesian_logistic_regression.py:
```diff
@@ -0,0 +1,365 @@
+        # pyplot.title('-ELBO against time for component {}'.format(t));
+        # pyplot.show()
+
+        pyplot.plot(range(len(guide_log_prob)), -1 * np.array(guide_log_prob), 'b-', label='- Guide log prob')
```
Generally want to separate plotting code from the algorithm itself.

gideonite commented on bbbvi.py:
```diff
@@ -0,0 +1,75 @@
+import matplotlib
+from matplotlib import pyplot
+
+def relbo(model, guide, *args, **kwargs):
```
Is this being used or is it being redefined elsewhere?

gideonite reviewed (changes requested):
Sorry for taking so long to get to this.

There are some code things to be cleaned up: mainly removing commented out code and refactoring plotting code.

More importantly, have we been able to factor out the relbo in a modular fashion?

lorenz-openai committed:
Back up before merging

lorenz-openai committed:
Merge branch 'wip/bimodal_experiments' of github.com:gideonite/projects2019-boosting-bbvi-pyro into wip/bimodal_experiments

sharrison5 reviewed (commented):
This is great, thanks!

A collection of fairly random comments on the notebook primarily:
 + Do we need to set `max_plate_nesting`?
 + Maybe stick with the `p`, `q` notation in the Variational Inference section, and stick with `s` for the boosting component?
 + Any mechanism for attribution etc - not clear at first glance if all the examples are contributed by the Pyro team?
 + Agree with Yu re the `data*10` statement - if we do need to keep it, perhaps just mention it as an aside?
Again, looks really good to me. Will let you know if I come across anything else, but this seems basically ready to try and submit AFAIC.

lorenz-openai committed:
Change variable naming, change data used

lorenzkuhn commented:
> I made some changes mostly to the text and committed them to a new branch (see #14). Below are two additional comments:

Thanks  for the review, @TNU-yaoy. Your changes to the notebook look good to me; I've merged them into this branch.

> 
> 1. you are using 3 different naming conventions for the latent and observed variables: z and x (intro), mu and y (model description), loc and obs (code). It might make sense to unify this if it is not too much work.

Good point, I've changed it to be z and x everywhere.

> 2. can we get away with not replicating the observations 10 times?

The approximation we get is quite a bit better if we use more data. Instead of replicating the list ten times, a cleaner way would be to draw some samples from N(4, 0.1) to get the observations. I've updated the notebook in this way, what do you think about that?

TNU-yaoy commented:
looks good to me

lorenz-openai committed:
Refoactor greedy algorithm section

lorenzkuhn commented:
> This is great, thanks!
> 
Hey @sharrison5, thanks a lot for having a look at the tutorial!

> A collection of fairly random comments on the notebook primarily:
> 
> * Do we need to set `max_plate_nesting`?

I think we don't strictly need to, i.e. Pyro will try to infer it if it's not explicitly given.

> * Maybe stick with the `p`, `q` notation in the Variational Inference section, and stick with `s` for the boosting component?

Makes sense, I've changed this.

> * Any mechanism for attribution etc - not clear at first glance if all the examples are contributed by the Pyro team?

Hmm, I didn't find any information on this on the Pyro repo or in the documentation. Maybe there's no such mechanism but that might be negotiable.

> * Agree with Yu re the `data*10` statement - if we do need to keep it, perhaps just mention it as an aside?

Fixed this by creating observations by sampling from N(4, 0.1).

>   Again, looks really good to me. Will let you know if I come across anything else, but this seems basically ready to try and submit AFAIC.

gideonite committed:
small notation / latex changes

gideonite committed:
more small latex changes

lorenz-openai committed:
Back up before pulling

lorenz-openai committed:
Add BLR with laplace guide
